source-container-build/app/source_build.py

```diff
@@ -528,6 +438,317 @@ def parse_image_name(image: str) -> tuple[str, str, str]:
+            new_layer = layer.save()
+            if new_layer != layer:
+                layer_descriptors[idx] = new_layer.descriptor
+                layer.delete()
+        return super().save()
+
+
+class Index:
+    """Represent the index.json of an OCI image"""
+
+    def __init__(self, layout: "OCIImage"):
+        self._layout = layout
+        self._content: IndexT | None = None
+        self._manifests: list[Manifest] | None = None
+
+    @property
+    def path(self) -> Path:
+        return self._layout.path / "index.json"
+
+    @property
+    def content(self) -> IndexT:
+        if self._content is None:
+            self._content = json.loads(self.path.read_text())
+        return self._content
+
+    def manifests(self) -> list[Manifest]:
+        """Return .manifests as a list of Manifest objects"""
+        if not self._manifests:
+            self._manifests = [Manifest(self._layout, d) for d in self.content["manifests"]]
+        return self._manifests
+
+    def save(self) -> None:
+        """Save this index"""
+        updated = False
+        for idx, manifest in enumerate(self.manifests()):
+            new_manifest = manifest.save()
+            if new_manifest != manifest:
+                self.content["manifests"][idx] = new_manifest.descriptor
+                updated = True
+        if updated:
+            self._manifests = None
+            self.path.write_text(json.dumps(self.content))
+
+
+class OCIImage:
+    """Represent an OCI image"""
+
+    def __init__(self, path: str | os.PathLike):
+        """Initialize this OCI image object
+
+        :param path: a path to an OCI image.
+        """
+        self._path = Path(path)
+        self._index: Index | None = None
+
+    @property
+    def path(self) -> Path:
+        return self._path
+
+    @property
+    def index(self) -> Index:
+        if self._index is None:
+            self._index = Index(self)
+        return self._index
+
+
+def merge_image(from_image: str, to_image: str) -> None:
+    """Merge sources from one to another
+
+    Image layers are appended to the destination image one by one.
+
+    :param from_image: str, merge sources from this image to another.
+    :param to_image: str, sources are merged into this image. Both of
+        ``from_image`` and ``to_image`` are directory paths holding sources in
+        OCI image layout format.
+    """
+
+    parent_image = OCIImage(from_image)
+    local_build = OCIImage(to_image)
+
+    parent_image_manifest = parent_image.index.manifests()[0]
+    local_build_manifest = local_build.index.manifests()[0]
+    for layer in parent_image_manifest.layers:
+        copy_dest_path = Path(local_build.path, "blobs", *layer.descriptor["digest"].split(":"))
+        shutil.copyfile(layer.path, copy_dest_path)
+        logger.debug("copy layer %s to %s", layer.path, copy_dest_path)
+        local_build_manifest.append_layer(layer)
```

Comment:
Doe we have any handling of duplicate layers? What happens if the local build has the same layer as the parent build?

Reply:
So far, the parent sources are the ones introduced by a OS parent image, and both the prefetched SRPMs and extra sources have their own names and generated layers. The original solution with BSI didn't solve duplicates as well.

STONEBLD-1831 I'm working on will handle the potential duplicate layers between parent sources and prefetched extra sources, i.e. SRPMs and dependencies.